lms/services/lti_outcomes.py

```diff
@@ -0,0 +1,235 @@
+from typing import NamedTuple
+from xml.etree import ElementTree
+
+from jinja2 import Template
+import requests
+from requests_oauthlib import OAuth1
+
+from lms.services.exceptions import ExternalRequestError, ServiceError
+
+
+__all__ = ["LTIOutcomesClient", "LTIOutcomesRequestParams"]
+
+LTI_OUTCOME_SERVICE_XML_NS = "http://www.imsglobal.org/services/ltiv1p1/xsd/imsoms_v1p0"
+
+# Envelope for LTI Outcome Service requests sent to the LMS
+#
+# See https://www.imsglobal.org/specs/ltiomv1p0/specification for specs for all
+# messages.
+#
+# nb. To be a valid XML document it is important that the template has no
+# whitespace at the start.
+LTI_OUTCOME_REQUEST_TEMPLATE = Template(
+    """<?xml version="1.0" encoding="UTF-8"?>
+<imsx_POXEnvelopeRequest xmlns="http://www.imsglobal.org/services/ltiv1p1/xsd/imsoms_v1p0">
+  <imsx_POXHeader>
+    <imsx_POXRequestHeaderInfo>
+      <imsx_version>V1.0</imsx_version>
+      <imsx_messageIdentifier>999999123</imsx_messageIdentifier>
+    </imsx_POXRequestHeaderInfo>
+  </imsx_POXHeader>
+  <imsx_POXBody>
+    {{ body }}
+  </imsx_POXBody>
+</imsx_POXEnvelopeRequest>
+"""
```

Comment:
Yes, although I have to say my experience in JS-land has been that since the introduction of template strings (the equivalent of multi-line f-strings) it has often been much more convenient from the point of view of comprehending everything to have templates inline when they are small. The downside is primarily with tooling (eg. syntax highlighting) that doesn't support multiple formats in one file, though that situation has improved a lot in the last couple of years.